### PR TITLE
JS: js/reflected-xss through file names

### PR DESCRIPTION
**Pull request by esben-semmle**
This PR makes js/reflected-xss treat file names as sources.

Other injection queries, such as js/sql-injection, can perhaps also use this new source, but I would like to introduce this source slowly to prevent too many FPs at once if it turns out that file names in practice are benign.

The PR is inspired by this vulnerability fix: <https://github.com/nunnly/m-server/commit/b8613fd1f35b2ba1e3b0cbac254c03f7c9f95232>, where a file name ultimately is concatenated with some HTML string fragments.

[Performance and results](<https://git.semmle.com/gist/esben/da8d5b0d705b824217d375a82334c808>) are unchanged, except for a result that I suspect is a FP due to a non-HTML response content-type that we fail to identify.

I have started on the 1.19 changes notes. I think we should aim for a less verbose library listing this time. I think a category and an example library is fine WDYT?

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/Concepts.qll [hunk 1]
```diff
@@ -29,6 +29,13 @@ abstract class FileSystemAccess extends DataFlow::Node {
   abstract DataFlow::Node getAPathArgument();
 }
 
+/**
+ * A data flow node that contains one or more file names from the local file system.
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/Concepts.qll]: Why "one or more"?

File: javascript/ql/src/semmle/javascript/security/dataflow/ReflectedXss.qll [hunk 2]
```diff
@@ -53,6 +53,13 @@ module ReflectedXss {
     }
   }
 
+  /** A source of file name input, considered as a flow source for reflected XSS. */
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/security/dataflow/ReflectedXss.qll]: The phrase "source of file name input" sounds quite awkward. Maybe just "a file name"?

**Comment by esben-semmle** [comment]
Comments addressed.

> OOI, why do you think the content-type in the newly identified result is non-HTML?

Several reasons:

- the file name is the only element in the response
- it is on a `/debug/` route
- the endpoint is used by other code in the same project, where it is embedded as an attribute value

**Comment by xiemaisi** [comment]
Query LGTM, but I'm having second thoughts about adding this as a source for reflected XSS. The alert message is phrased in terms of user-provided values, and I don't really think local file system paths fit into that category. Is this perhaps a separate query? (Ping @asger-semmle for input.)

**Comment by asger-semmle** [comment]
I don't understand the vulnerability you want to flag here?

As far as I can tell, the commit doesn't actually fix anything. It's a static file server, so attackers have no control over file names.

If they could upload files and thereby control file names, I'd categorize it as stored XSS.

**Comment by esben-semmle** [comment]
> I don't understand the vulnerability you want to flag here?

It is a vulnerability, no doubt about that ([hackerone](https://hackerone.com/reports/319794) agrees).
It is dangerous to rely on the format of strings that the application does not control at all.

> If they could upload files and thereby control file names, 

They do not have to choose the file names through the application with the XSS vulnerability. Imagine an escalation through a compromised FTP server with a directory that happens to be served by  `m-server`...

> I'd categorize it as stored XSS.

Yes, the source should perhaps be moved to XSS.ql, I had a wrong definition in mind. Reflected XSS is not stored, by definition:

 [OWASP](https://www.owasp.org/index.php/Cross-site_Scripting_(XSS)#Reflected_XSS_Attacks):
> Reflected XSS is also sometimes referred to as Non-Persistent or Type-II XSS

**Comment by xiemaisi** [comment]
I like Asger's suggestion of treating this as a stored XSS. We don't have a query for that yet (merging into `XSS.ql` doesn't seem right; that query is for DOM-based XSS), but adding one would increase our claimed security coverage, which is a good thing.

**Comment by esben-semmle** [comment]
OK. Closing for now then.

File: javascript/ql/src/semmle/javascript/frameworks/NodeJSLib.qll [hunk 3]
```diff
@@ -356,6 +365,22 @@ module NodeJSLib {
...
 
+  /**
+   * A data flow node that contains one or more file names from the local file system.
+   */
+  private class NodeJSFileNameSource extends FileNameSource {
+
+    NodeJSFileNameSource() {
+      exists (string name |
+        name = "readdir" or
+        name = "realpath" |
```
[the remark this answers is not in the document]
Comment (esben-semmle) [reply javascript/ql/src/semmle/javascript/frameworks/NodeJSLib.qll]: Yes,  I believe so, but I still think it should remain a source. 
If we later find that `readdir` or `realpath` cause many FPs as sources, we can downgrade them to taint steps.

**Pull request closed without merging**